core/src/plan/primary_key.rs

```diff
@@ -101,113 +195,407 @@ impl<'a> PrimaryKeyPlanner<'a> {
     fn expr(
         &self,
         outer_context: Option<Rc<Context<'a>>>,
-        current_context: Option<Rc<Context<'a>>>,
+        current_context: &Rc<Context<'a>>,
         expr: Expr,
     ) -> PrimaryKey {
-        let check_primary_key = |key: &Expr| {
-            let key = match key {
-                Expr::Identifier(ident) => ident,
-                Expr::CompoundIdentifier { ident, .. } => ident,
-                _ => return false,
+        enum PossibleResults {
+            Matched(PrimaryKey),
+            Retry(Box<Expr>, Box<Expr>),
+        }
+
+        // Returns the primary key variant associated to the provided key and value, if any.
+        let get_primary_key = |key: Box<Expr>, value: Box<Expr>| {
+            if !(check_evaluable(Some(current_context.clone()), &key)
+                && check_evaluable(None, &value))
+            {
+                return PossibleResults::Retry(key, value);
```

Comment:
I am also not sure that we really need `PossibleResults::Retry` like repeatable logics.
the current state of primary key planner is very simple. just a logical planner.
if the planner find the first matched one, then it's enough. so there is no logic required for selecting the best options or something like that.

Reply:
The Retry is needed to avoid some avoidable clones. If you prefer to clone, I can remove it.

Reply:
See commit 373d427

Reply:
ok, if that `Retry` is being used because of avoiding clones, then let's use clones. then we can find the way to remove clone without Retry logics.

Reply:
Removed Retry. I do not think we can remove the clones without introducing the Retry variant.